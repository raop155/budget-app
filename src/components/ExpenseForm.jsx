import React, { useState, useEffect } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export const ExpenseForm = (props) => {
  const [description, setDescription] = useState(props.expense ? props.expense.description : '');
  const [amount, setAmount] = useState(props.expense ? (props.expense.amount / 100).toString() : 0);
  const [createAt, setCreateAt] = useState(
    props.expense ? moment(props.expense.createAt) : moment(),
  );
  const [note, setNote] = useState(props.expense ? props.expense.note : '');

  const [error, setError] = useState({});
  const [focused, setFocused] = useState(false);

  useEffect(() => {
    console.log('useEffect');
  }, []);

  const onChangeHandler = (e) => {
    const { name, value } = e.target;

    if (name === 'description') {
      setDescription(value);
    } else if (name === 'amount') {
      if (!value || /^\d{1,}(\.\d{0,2})?$/.test(value)) setAmount(value);
    } else if (name === 'note') {
      setNote(value);
    }
  };

  const onDateChange = (createAt) => {
    if (createAt) setCreateAt(createAt);
  };

  const onFocusChange = ({ focused }) => {
    setFocused(focused);
  };

  const onSubmit = (e) => {
    e.preventDefault();

    if (!description || !amount) {
      setError({
        form: 'Please provide description and amount.',
      });
    } else {
      setError({
        form: '',
      });

      props.onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createAt: createAt.valueOf(),
        note,
      });
    }
  };

  return (
    <div>
      {error.form && <p className='form__error'>{error.form}</p>}
      <form className='form' onSubmit={onSubmit}>
        <label className='label' htmlFor='description'>
          Description
        </label>
        <input
          id='description'
          className='text-input'
          type='text'
          name='description'
          placeholder='Ex: Coffee'
          autoFocus
          value={description}
          onChange={onChangeHandler}
        />
        <label className='label' htmlFor='amount'>
          Amount
        </label>
        <input
          id='amount'
          className='text-input'
          type='text'
          name='amount'
          placeholder='Ex: 2.99'
          value={amount}
          onChange={onChangeHandler}
        />
        <label className='label' htmlFor='expense-form-dp'>
          Date
        </label>
        <SingleDatePicker
          date={createAt}
          onDateChange={onDateChange}
          focused={focused}
          onFocusChange={onFocusChange}
          id='expense-form-dp'
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <label className='label'>Note</label>
        <textarea
          className='textarea'
          name='note'
          placeholder='Add a note htmlFor your expense (optional)'
          value={note}
          onChange={onChangeHandler}
        ></textarea>
        <div>
          <button className='button'> {props.expense ? 'Edit' : 'Add'} Expense</button>
        </div>
      </form>
    </div>
  );
};

export default ExpenseForm;
