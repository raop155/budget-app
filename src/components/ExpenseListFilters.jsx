import React, { useState } from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

const ExpenseListFilters = props => {
  const [focused, setFocused] = useState(null);

  const onDatesChange = ({ startDate, endDate }) => {
    props.dispatch(setStartDate(startDate));
    props.dispatch(setEndDate(endDate));
  };

  const onFocusChange = focused => {
    setFocused(focused);
  };

  return (
    <>
      <input
        type='text'
        value={props.filters.text}
        onChange={e => {
          props.dispatch(setTextFilter(e.target.value));
        }}
      />
      <select
        value={props.filters.sortBy}
        onChange={e => {
          if (e.target.value === 'date') {
            props.dispatch(sortByDate());
          } else if (e.target.value === 'amount') {
            props.dispatch(sortByAmount());
          }
        }}
      >
        <option value='date'>Date</option>
        <option value='amount'>Amount</option>
      </select>

      <DateRangePicker
        startDate={props.filters.startDate}
        endDate={props.filters.endDate}
        onDatesChange={onDatesChange}
        focusedInput={focused}
        onFocusChange={onFocusChange}
        startDateId='expense-list-filters-dp-start'
        endDateId='expense-list-filters-dp-end'
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
    </>
  );
};

const mapStateToProps = state => {
  return {
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
