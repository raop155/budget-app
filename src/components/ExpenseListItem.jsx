import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItem = ({ id, description, amount, createAt }) => {
  return (
    <div className='content-container'>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>
        amount: {numeral(amount / 100).format('$0,0.00')} - createAt:{' '}
        {moment(createAt).format('MMMM Do, YYYY')}
      </p>
    </div>
  );
};

export default ExpenseListItem;
