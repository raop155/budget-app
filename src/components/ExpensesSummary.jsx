import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
  const { expenseCount } = props;
  let { expensesTotal } = props;
  const expenseWord = expenseCount > 1 ? 'expenses' : 'expense';
  expensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <h1>
      Viewing {expenseCount} {expenseWord} totalling {expensesTotal}
    </h1>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
