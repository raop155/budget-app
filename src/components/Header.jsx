import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => {
  return (
    <header>
      <h1>Heroku Budget App</h1>
      <nav>
        <NavLink to='/' activeClassName='is-active' exact>
          Dashboard
        </NavLink>
        <NavLink to='/create' activeClassName='is-active'>
          Create Expense
        </NavLink>
        <NavLink to='/help' activeClassName='is-active'>
          Help
        </NavLink>
      </nav>
    </header>
  );
};

export default Header;
