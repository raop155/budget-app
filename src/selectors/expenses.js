import moment from 'moment';

const selectExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      const createAt = moment(expense.createAt);
      const startDateMatch = startDate ? startDate.isSameOrBefore(createAt, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(createAt, 'day') : true;
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createAt < b.createAt ? 1 : -1;
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      } else {
        return 1;
      }
    });
};

export default selectExpenses;
