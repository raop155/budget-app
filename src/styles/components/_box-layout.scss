.box-layout {
  display: flex;
  align-items: center;
  justify-content: center;
  background: url('/images/bg.jpg');
  background-size: cover;
  height: 100vh;
  width: 100vw;
}

.box-layout__box {
  background: fade-out(white, 0.15);
  border-radius: 5px;
  text-align: center;
  width: 300px;
  padding: $l-size $m-size;
}

.box-layout__title {
  margin: 0 0 $m-size 0;
  line-height: 1;
}
