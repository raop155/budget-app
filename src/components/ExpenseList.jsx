import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { startSetExpenses } from '../actions/expenses';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => {
  const { startSetExpenses, expenses } = props;

  useEffect(() => {
    console.log('useEffect ExpenseList');
    startSetExpenses();
  }, [startSetExpenses]);

  return (
    <div className='content-container'>
      <div className='list-header'>
        <div>Expense</div>
        <div className='show-desktop'>Amount</div>
      </div>
      <div className='list-body'>
        {expenses.length === 0 ? (
          <div className='list-item list-item--message'>
            <span>No expenses</span>
          </div>
        ) : (
          expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />;
          })
        )}
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    startSetExpenses: () => dispatch(startSetExpenses()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseList);
