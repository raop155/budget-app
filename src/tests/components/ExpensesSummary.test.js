import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

describe('<ExpensesSummary /> should', () => {
  it('render correctly', () => {
    const wrapper = shallow(
      <ExpensesSummary expenses={expenses} expensesTotal={expenses.length} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
