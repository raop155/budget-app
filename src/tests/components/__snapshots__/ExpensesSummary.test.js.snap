// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpensesSummary /> should render correctly 1`] = `
<h1>
  Viewing 
   
  expense
   totalling 
  $0.03
</h1>
`;
